base/generator.jl

```diff
@@ -126,3 +126,4 @@ IteratorEltype(::Type) = HasEltype()  # HasEltype is the default
 IteratorEltype(::Type{Generator{I,T}}) where {I,T} = EltypeUnknown()
 
 IteratorEltype(::Type{Any}) = EltypeUnknown()
+IteratorEltype(::Type{Union{}}) = EltypeUnknown()
```

Comment:
Should this perhaps have known element type `Union{}`? I could imagine this pessimizing something.

Reply:
`eltype(Union{}})` throws an error, so I think this is fine.

Reply:
This is unfortunate in combination with
https://github.com/JuliaLang/julia/blob/1968b239fa6f5ab266e62e2240d828f37622ee83/base/generator.jl#L123
If `x` cannot be inferred exactly, we get `typeof(x)::Type{<:Something}`, which includes `Type{Union{}}`, although `x` will of course never be `::Union{}`. E.g.
```julia
julia> code_warntype(Base.IteratorEltype, Tuple{Array{Int}})
Body::Union{EltypeUnknown, HasEltype}
123 1 ─ %1 = (Base.typeof)(x)::Type{#s55} where #s55<:Array{Int64,N} where N                                    │
    │   %2 = (isa)(%1, Type{Union{}})::Bool                                                                     │
    └──      goto #3 if not %2                                                                                  │
    2 ─      goto #4                                                                                            │
    3 ─ %5 = (Base.IteratorEltype)(%1)::Union{EltypeUnknown, HasEltype}                                         │
    └──      goto #4                                                                                            │
    4 ┄ %7 = φ (#2 => $(QuoteNode(Base.EltypeUnknown())), #3 => %5)::Union{EltypeUnknown, HasEltype}            │
    └──      return %7   
```
Compare with this method removed:
```julia
julia> code_warntype(Base.IteratorEltype, Tuple{Array{Int}})
Body::Base.HasEltype
123 1 ─     return $(QuoteNode(Base.HasEltype()))  
```
This causes the regression noted in #29464.

Reply:
fwiw
```
julia> code_warntype(Base.IteratorEltype, Tuple{Array{Int}})
Body::Base.HasEltype
123 1 ─     return $(QuoteNode(Base.HasEltype()))

julia> versioninfo()
Julia Version 1.0.0
Commit 5d4eaca0c9 (2018-08-08 20:58 UTC)
OS: Linux (x86_64-pc-linux-gnu)
LLVM: libLLVM-6.0.0 (ORCJIT, skylake)

julia> code_warntype(Base.IteratorEltype, Tuple{Array{Int}})
Body::Base.HasEltype
123 1 ─     return $(QuoteNode(Base.HasEltype()))                                           │
```

Reply:
Ah, that won't quite work. Instead we can add a special case for flatten when `eltype(I)` is `Union{}`.